Integrations: AWS OIDC - ListDatabases action

marcoandredinis opened the pull request:
This PR adds a new Action for the AWSOIDC Integration: ListDatabases
The goal of this action is to provide the User a list of RDS Databases
from which the User will pick one to be added as a Teleport Database
Resource.
This way, the user doesn't need to copy/paste DB name, endpoints and
labels.

Before being able to call this action, the User has to set up an AWS
OIDC integration.

How does it work:
```
                    Client (web app)
                     │     ▲
                     │     │4. Returns list of DBs
 1. List Databases   │     │   (name, tags, status, endpoint)
                     │     │
                     │     │
                     │     │
                     │     │
                     ▼     │       3. rds.DescribeDBInstances
               ┌───────────┴────┐     (auth: token)                ┌─────────┐
               │                ├─────────────────────────────────►│         │
               │ Teleport Proxy │                                  │   AWS   │
               │                │     3.1. Get OIDC Config         │         │
               │                │◄─────────────────────────────────┤         │
               │                │     3.2. Get RSA Public Key      │         │
               │                │◄─────────────────────────────────┤         ├─────────┐ 3.3.
               │                │                                  │         │         │Validates token signature
               │                │                                  │         │         │with received public key
               │                │   3.4 Returns list of DBs        │         │◄────────┘
               │                │◄─────────────────────────────────┤         │
               └─┬──────────────┘                                  └─────────┘
                 │
                 │  2. Sign Token
                 │
                 ▼
              ┌───────────────────┐
              │                   │
              │   Teleport Auth   │
              │  RSA Private Key  │
              └───────────────────┘

```

Eg of an HTTP response from `4.`
```json
{
   "databases":[
      {
         "name":"marcodb-test",
         "desc":"RDS instance in us-east-1",
         "protocol":"postgres",
         "type":"rds",
         "labels":[
            {
               "name":"account-id",
               "value":"123456789012"
            },
            {
               "name":"canBeDeleted",
               "value":"yes"
            },
            {
               "name":"endpoint-type",
               "value":"instance"
            },
            {
               "name":"engine",
               "value":"postgres"
            },
            {
               "name":"engine-version",
               "value":"14.6"
            },
            {
               "name":"env",
               "value":"test"
            },
            {
               "name":"project",
               "value":"discover"
            },
            {
               "name":"region",
               "value":"us-east-1"
            },
            {
               "name":"status",
               "value":"available"
            },
            {
               "name":"teleport.dev/origin",
               "value":"cloud"
            }
         ],
         "hostname":"marcodb-test.abcdef123456.us-east-1.rds.amazonaws.com",
         "uri":"marcodb-test.abcdef123456.us-east-1.rds.amazonaws.com:5432",
         "aws":{
            "region":"us-east-1",
            "rds":{
               "instance_id":"marcodb-test",
               "resource_id":"db-123456ABCEDF123456ABCEDF00",
               "iam_auth":false
            },
            "account_id":"123456789012",
            "status":"available"
         }
      }
   ]
}
```

kimlisa commented on lib/web/ui/integration.go:
```diff
@@ -102,3 +103,73 @@ func MakeIntegration(ig types.Integration) Integration {
 }
+
+// AWSOIDCListDatabasesRequest is a request to ListDatabases using the AWS OIDC Integration.
+type AWSOIDCListDatabasesRequest struct {
```
i think i saw labels in the response too right? `ui.Label` format?

kimlisa commented on lib/web/ui/integration.go:
```diff
@@ -102,3 +103,73 @@ func MakeIntegration(ig types.Integration) Integration {
+
+	// NextToken is used for pagination.
+	// If non-empty, it can be used to request the next page.
+	NextToken string `json:"next_token,omitempty"`
```
camel?
```suggestion
	NextToken string `json:"nextToken,omitempty"`
```

marcoandredinis replied:
Yes 👍

marcoandredinis committed:
AWSOIDC Integration: ListDatabases

This PR adds a new Action for the AWSOIDC Integration: ListDatabases
The goal of this action is to provide the User a list of RDS Databases
from which the User will pick one to be added as a Teleport Database
Resource.
This way, the user doesn't need to copy/paste DB name, endpoints and
labels.

Before being able to call this action, the User has to set up an AWS
OIDC integration.

How does it work:
```
                    Client (web app)
                     │     ▲
                     │     │4. Returns list of DBs
 1. List Databases   │     │   (name, tags, status, endpoint)
                     │     │
                     │     │
                     │     │
                     │     │
                     ▼     │       3. rds.DescribeDBInstances
               ┌───────────┴────┐     (auth: token)                ┌─────────┐
               │                ├─────────────────────────────────►│         │
               │ Teleport Proxy │                                  │   AWS   │
               │                │     3.1. Get OIDC Config         │         │
               │                │◄─────────────────────────────────┤         │
               │                │     3.2. Get RSA Public Key      │         │
               │                │◄─────────────────────────────────┤         ├─────────┐ 3.3.
               │                │                                  │         │         │Validates token signature
               │                │                                  │         │         │with received public key
               │                │   3.4 Returns list of DBs        │         │◄────────┘
               │                │◄─────────────────────────────────┤         │
               └─┬──────────────┘                                  └─────────┘
                 │
                 │  2. Sign Token
                 │
                 ▼
              ┌───────────────────┐
              │                   │
              │   Teleport Auth   │
              │  RSA Private Key  │
              └───────────────────┘

```

marcoandredinis commented:
@kimlisa @r0mant I can split up the PR
It's not that much production code, but comments, tests and proto auto generated code, leads to a lot of changes

marcoandredinis committed:
add resource and account ids to DB resource

r0mant commented on lib/integrations/awsoidc/listdatabases.go:
```diff
@@ -0,0 +1,314 @@
+
+// convertDBClusterToDatabase converts a Database from its rds/types.DBCluster representation into
+// a Teleport representation.
+func convertDBClusterToDatabase(in types.DBCluster) (*AWSDatabase, error) {
```
Same question as above, can we just reuse existing "db cluster to database resource" we use in auto-discovery.

marcoandredinis replied:
See comment above

r0mant commented on lib/web/ui/integration.go:
```diff
@@ -102,3 +103,67 @@ func MakeIntegration(ig types.Integration) Integration {
+}
+
+// AWSDatabase is a representation of an AWS RDS Database
+type AWSDatabase struct {
```
Similar question as above about these new resources, can web UI use exiting database types we already have defined? cc @kimlisa Just trying to avoid having too many representations of the same thing.

kimlisa replied:
i think it makes sense to me to be explicit like this, a type that tells me that this isn't a teleport database resource

that said, we can re-use the existing web/ui database type (linked below) and extend it to include fields unique to the aws database if we really want to (in the end it doesn't matter much as long as the front get the values we need b/c the frontend will customize it further to suite the needs of the component that needs it anyways):
https://github.com/gravitational/teleport/blob/e2df4a31ba8f810e266180f5bcba451f01bf2397/lib/web/ui/server.go#L227

no strong opinions

marcoandredinis replied:
I've added all the types.AWS fields and the Status as well (which comes from a label).

r0mant reviewed (commented):
Nice job on the test coverage here 👍 

Left a few suggestions on reusing existing resource representations and simplifying the API a bit.

marcoandredinis commented on lib/web/integrations.go:
```diff
@@ -181,3 +182,44 @@ func (h *Handler) integrationsList(w http.ResponseWriter, r *http.Request, p htt
 	}, nil
 }
+
+func (h *Handler) integrationsExecute(w http.ResponseWriter, r *http.Request, p httprouter.Params, sctx *SessionContext, site reversetunnel.RemoteSite) (interface{}, error) {
```
Sure :+1: 
I'll create one endpoint per action.
So in this case we'll have
```go
h.POST("/webapi/sites/:site/integrations/:name/action/list_databases", h.WithClusterAuth(h.awsOIDCListDatabases))
```

r0mant replied:
Why do we need "action" at all? Could we just do

```
/webapi/sites/:site/integrations/aws/:name/databases
```

Btw, I'd also add /aws to the path to be able to differentiate between handlers for different integrations.

marcoandredinis replied:
Looks good :+1: 
I've just replace `aws` with `aws-oidc` because that's the name we are using everywhere

marcoandredinis committed:
move api namespaces

marcoandredinis commented on lib/integrations/awsoidc/listdatabases.go:
```diff
@@ -0,0 +1,314 @@
+
+// convertDBInstanceToDatabase converts a Database from its rds/types.DBInstance representation into
+// a Teleport representation.
+func convertDBInstanceToDatabase(in types.DBInstance) (*AWSDatabase, error) {
```
We are using aws go sdk v1 for rds api everywhere else
https://github.com/gravitational/teleport/blob/05aa0488a178292cfa384b491589d54934d4c76b/lib/cloud/aws/tags_helpers.go#L23
I don't think I can re-use such helpers

I decided to use v2, so most things will not be compatible.
Now, I'm wondering if I should switch to the v1 of the go sdk :thinking:

marcoandredinis replied:
I've created new helpers using the aws sdk go v2 :+1:

marcoandredinis committed:
use types.Database instead of custom database format

marcoandredinis committed:
add database uri

kimlisa commented on lib/web/ui/server.go:
```diff
@@ -238,17 +238,35 @@ type Database struct {
+type AWS struct {
+	// embeds types.AWS fields into this struct when des/serializing.
+	types.AWS `json:""`
+	// Status describes the current server status as reported by the
+	// Currently this field is populated for AWS RDS Databases when Listing Databases using the AWS OIDC Integration
```
incomplete comment

marcoandredinis commented on api/types/integration_awsoidc.go:
```diff
@@ -0,0 +1,43 @@
+	IntegrationAWSOIDCSubject = "system:proxy"
+)
+
+// GenerateAWSOIDCTokenRequest are the parameters used to request an AWS OIDC Integration token.
+type GenerateAWSOIDCTokenRequest struct {
+	// Issuer is the entity that is signing the JWT.
+	// This value must contain the AWS OIDC Integration configured provider (Proxy's Public URL).
+	Issuer string `json:"issuer"`
+}
```
I did it only to implement the `CheckAndSetDefaults`
The pb type is created here: `api/gen/proto/go/teleport/integration/v1/integration_service.pb.go`
I was trying to avoid adding custom code close in those auto generated folders.

marcoandredinis commented on lib/auth/integration/integrationv1/service_test.go:
```diff
@@ -382,3 +401,49 @@ func initSvc(t *testing.T, kind string) (context.Context, localClient, *Service)
+	keystore   *keystore.Manager
+}
+
+// CAGetter describes the required methods to sign a JWT to be used for AWS OIDC Integration.
```
Yes! Thank you.

marcoandredinis commented on lib/web/apiserver.go:
```diff
@@ -710,6 +710,15 @@ func (h *Handler) bindDefaultEndpoints() {
 
+	h.POST("/webapi/sites/:site/integrations/aws-oidc/:name/databases", h.WithClusterAuth(h.awsOIDCListDatabases))
+
+	// AWS OIDC Integration specific endpoints:
+	// Unauthenticated access to OpenID Configuration - used for AWS OIDC IdP integration
+	// h.GET("/.well-known/openid-configuration", h.WithLimiter(h.openidConfiguration))
+	// h.GET(OIDCJWKWURI, h.WithLimiter(h.jwksOIDC))
```
Oops, I meant to add this with the limiter as they are public endpoints

marcoandredinis committed:
fix comments and rate limiter

marcoandredinis committed:
test name override when converting RDS V2 DBs

marcoandredinis committed:
fix webapi database URI field

marcoandredinis committed:
TestClusterDatabasesGet: add parallel
